### PR TITLE
fix: concurrency stuff on upload repo

### PR DESCRIPTION
**Pull request by biancarosa**

### CODE REVIEW

**Review by maxamillion — APPROVED** [review]

File: pkg/services/files/uploader.go [hunk 1]
```diff
@@ -101,26 +97,29 @@ type uploadDetails struct {
...
 //If the error "Too many open files" appears, lower the number of workers
-func createWorkers() {
+func createWorkers(uploadQueue chan *uploadDetails) {
 	if !workersCreated {
 		numberOfWorkers := 100
 		for i := 0; i < numberOfWorkers; i++ {
-			go worker()
+			go worker(uploadQueue)
 		}
 		workersCreated = true
```
Comment (AaronH88) [inline pkg/services/files/uploader.go:124]: Have you testing this locally?
I think this bool will prevent more than one repo to be uploaded, have you tried creating two or three upload requests?

This whole func can be removed, and creating 100 workers can fall back into the UploadRepo() func

File: pkg/services/files/uploader.go [hunk 2]
```diff
@@ -19,9 +18,6 @@ import (
 //Bool used to control if uploading workers need to be created
 var workersCreated = false
```
Comment (AaronH88) [inline pkg/services/files/uploader.go:19]: I think this bool is no longer needed

File: pkg/services/files/uploader.go [hunk 3]
```diff
@@ -153,12 +155,21 @@ func (u *S3Uploader) UploadRepo(src string, account string) (string, error) {
...
+		res.done = make(chan bool)
+		uploadQueue <- res
+		uploadDetailsList = append(uploadDetailsList, res)
+		count++
 		return nil
 	})
-
-	repoWaitGroup.Wait()
+	log.Infof("Files are being uploaded.... %d files to upload", len(uploadDetailsList))
+	for i, u := range uploadDetailsList {
```
Comment (AaronH88) [inline pkg/services/files/uploader.go:166]: "u" is already defined in this function, can we rename this var please? I got confused for a min wondering why we are looking for "uploader"

Comment (AaronH88) [reply pkg/services/files/uploader.go:166]: Do we want to do some sort of conditional to ensure that i is equal to count (-1) at the end of this loop?

**Review by AaronH88 — CHANGES_REQUESTED** [review]
I think the bool "workersCreated" needs to be removed, and some small var name changes.

Apart from that, it looks good to me :smile:

File: pkg/services/files/uploader.go (hunk 1 above)
Comment (biancarosa) [reply pkg/services/files/uploader.go:124]: I tested but just with one request. I can see how the bool will be a problem and I'm removing it :)

File: pkg/services/files/uploader.go [hunk 4]
```diff
@@ -134,13 +118,20 @@ func (u *S3Uploader) UploadRepo(src string, account string) (string, error) {
 	log.Debugf("S3Uploader::UploadRepo::src: %#v", src)
 	log.Debugf("S3Uploader::UploadRepo::account: %#v", account)
 
-	createWorkers()
+	uploadQueue := make(chan *uploadDetails)
+
+	numberOfWorkers := cfg.UploadWorkers
```
Comment (AaronH88) [inline pkg/services/files/uploader.go:123]: NICE!

**Comment by AaronH88** [comment]
Looks good to me, lets give it a go and see what happens

**Review by AaronH88 — APPROVED** [review]
Great!

**Comment by biancarosa** [comment]
I'm thinking IQE is not happy with some latest changes so I'm gonna merge this and see what happens.

**Pull request merged**